### PR TITLE
Enhance xml parser, xml files to generate enum list and index directly

### PR DESCRIPTION
**Pull request by subhasisb**
#### Describe Bug or Feature

1. Master XML file parser code was too old and bad
2. Master XML's were parsed to generate [job/node/svr/sched/queue]_attr_def.c but correspoding enum list for the indexes are hard coded. This means any change to the order in the master xml files will not reflect in the enum list and will introduce errors silently (won't fail compilation
3. There is no enum list for the resc_def_all.c. find_resc_def() iterates through every resource doing string matching. This is absolutely unnecessary for the cases where the index is known upfront.

#### Describe Your Change

1. Simplified the attr_parser.py code
2. Updated the XML input files to have information about the member_index
3. Enhanced the parser and to use the member_index value for each attribute to automate the enum lists [job/node/svr/sched/queue]_attr_def.h
4. In cases where the resource name is known upfront, changed calls to find_resc_def() to directly index the resc array with the index (which is now available)

#### Attach Test and Valgrind Logs/Output

Description: Tests from given sources on platforms CENTOS7, UBUNTU1804, SUSE15
Platforms: CENTOS7,UBUNTU1804,SUSE15
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|2456|1794|0|0|1|0|1793|

Description: Tests from given sources on platforms CENTOS7, UBUNTU1804, SUSE15, UBUNTU1604
Platforms: CENTOS7,UBUNTU1804,SUSE15,UBUNTU1604
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|2461|3510|0|0|0|0|3510|

valgrind run logs: 
[valgrind.log](https://github.com/openpbs/openpbs/files/4891276/valgrind.log)

### CODE REVIEW

File: src/include/Makefile.am [hunk 1]
```diff
@@ -126,8 +126,56 @@ noinst_HEADERS = \
...
+	resv_attr_enum.h \
+	sched_attr_enum.h \
+	svr_attr_enum.h \
 	svrfunc.h \
 	ticket.h \
 	tracking.h \
 	user.h \
 	work_task.h
+	win.h \
+	win_remote_shell.h \
```
Comment (bhroam) [inline src/include/Makefile.am:142]: Aren't these going away?

File: src/resmom/linux/mom_start.c [hunk 2]
```diff
@@ -233,8 +233,7 @@ getplacesharing(job *pjob)
 	 *	Compute the "Resource_List.place" index for vnss[][]:
 	 */
 	if (prsdef == NULL) {
-		prsdef = find_resc_def(svr_resc_def, "place",
-			svr_resc_size);
+		prsdef = &svr_resc_def[RESC_PLACE];
```
Comment (bhroam) [inline src/resmom/linux/mom_start.c:236]: No need for the braces

File: src/resmom/mock_run.c [hunk 3]
```diff
@@ -181,14 +181,14 @@ mock_run_mom_set_use(job *pjob)
...
-		rd[3] = find_resc_def(svr_resc_def, "cpupercent", svr_resc_size);
+		rd[0] = &svr_resc_def[RESC_NCPUS];
+		rd[1] = &svr_resc_def[RESC_MEM];
+		rd[2] = &svr_resc_def[RESC_CPUT];
+		rd[3] = &svr_resc_def[RESC_CPUPERCENT];
 		rd[4] = NULL;
 	}
 	if (vmemd == NULL) {
-		vmemd = find_resc_def(svr_resc_def, "vmem", svr_resc_size);
+		vmemd = &svr_resc_def[RESC_VMEM];
```
Comment (bhroam) [inline src/resmom/mock_run.c:191]: No need for the assert, and then no need for the braces.

File: src/server/svr_func.c [hunk 4]
```diff
@@ -1653,12 +1653,12 @@ static int svr_oldstyle[] = {
 	(int)SRV_ATR_max_running,
 	(int)SRV_ATR_MaxUserRun,
 	(int)SRV_ATR_MaxGrpRun,
-	(int)SVR_ATR_MaxUserRes,
-	(int)SVR_ATR_MaxGroupRes,
-	(int)SVR_ATR_MaxUserRunSoft,
-	(int)SVR_ATR_MaxGrpRunSoft,
-	(int)SVR_ATR_MaxUserResSoft,
-	(int)SVR_ATR_MaxGroupResSoft,
+	(int)SRV_ATR_MaxUserRes,
```
Comment (bhroam) [inline src/server/svr_func.c:1656]: Why the change from SVR to SRV?

Comment (subhasisb) [reply src/server/svr_func.c:1656]: Half the variables are prepended with SVR and half are SRV, so went one way..

File: src/include/Makefile.am (hunk 1 above)
Comment (subhasisb) [reply src/include/Makefile.am:142]: yes somehow happened after the rebase I guess, fixed

File: src/server/svr_func.c (hunk 4 above)
Comment (bhroam) [reply src/server/svr_func.c:1656]: Ahh, yes consistency is good.  I personally would have voted for SVR, but I guess it doesn't really matter.  What is the rest of the code doing (function names and the like)?

File: src/lib/Libattr/master_sched_attr_def.xml [hunk 5]
```diff
@@ -95,449 +95,471 @@
...
+	<member_at_type>ATR_TYPE_LONG</member_at_type>
 	<member_at_parent>PARENT_TYPE_SCHED</member_at_parent>
 	<member_verify_function>
 	<ECL>verify_datatype_time</ECL>
 	<ECL>NULL_VERIFY_VALUE_FUNC</ECL>
 	</member_verify_function>
    </attributes>
-   <attributes>  /* do_not_span_psets */
-	<member_name><both>ATTR_do_not_span_psets</both></member_name> 	<!-- do_not_span_psets -->
+   <attributes>  <member_index>SCHED_ATR_dont_span_psets</member_index>
```
Comment (prakashcv13) [inline src/lib/Libattr/master_sched_attr_def.xml:149]: <member_index> should move to the next line.

File: src/lib/Libattr/master_sched_attr_def.xml [hunk 6]
```diff
@@ -95,449 +95,471 @@
...
 	<member_at_parent>PARENT_TYPE_SCHED</member_at_parent>
 	<member_verify_function>
 	<ECL>verify_datatype_bool</ECL>
 	<ECL>NULL_VERIFY_VALUE_FUNC</ECL>
 	</member_verify_function>
    </attributes>
-   <attributes> /* job_run_wait */
-        <member_name><both>ATTR_job_run_wait</both></member_name>    <!-- job_run_wait -->
+   <attributes>
+	<member_index>SCHED_ATR_job_run_wait</member_index>
```
Comment (prakashcv13) [inline src/lib/Libattr/master_sched_attr_def.xml:247]: alignment.

File: src/lib/Libattr/master_sched_attr_def.xml [hunk 7]
```diff
@@ -95,449 +95,471 @@
...
         <member_verify_function>
         <ECL>verify_datatype_long</ECL>
         <ECL>NULL_VERIFY_VALUE_FUNC</ECL>
         </member_verify_function>
    </attributes>
-   <attributes>   /* job_sort_formula */
-    <member_name><both>ATTR_job_sort_formula</both></member_name>     <!-- "job_sort_formula" -->
+   <attributes>
+	<member_index>SCHED_ATR_job_sort_formula</member_index>
+    <member_name>ATTR_job_sort_formula</member_name>     <!-- "job_sort_formula" -->
```
Comment (prakashcv13) [inline src/lib/Libattr/master_sched_attr_def.xml:520]: under indent, this and subsequent lines.

File: src/lib/Libattr/master_svr_attr_def.xml [hunk 8]
```diff
@@ -92,1715 +92,1717 @@
...
 	<member_at_parent>PARENT_TYPE_SERVER</member_at_parent>
 	<member_verify_function>
 	<ECL>verify_datatype_long</ECL>
 	<ECL>NULL_VERIFY_VALUE_FUNC</ECL>
 	</member_verify_function>
    </attributes>
    <attributes>
-   /* SRV_ATR_MaxGrpRun */
-	<member_name><both>ATTR_maxgrprun</both></member_name>		<!-- "max_group_run" -->
+    <member_index>SRV_ATR_MaxGrpRun</member_index>
```
Comment (prakashcv13) [inline src/lib/Libattr/master_svr_attr_def.xml:282]: alignment

File: buildutils/attr_parser.py [hunk 9]
```diff
@@ -90,762 +94,191 @@ def match(self, *args):
             return False
 
 
-def fileappend(line):
+def fileappend(prop_type, line):
     """
```
Comment (prakashcv13) [inline buildutils/attr_parser.py:98]: let us add a docstring describing the parameters

File: buildutils/attr_parser.py [hunk 10]
```diff
@@ -90,762 +94,191 @@ def match(self, *args):
...
+    elif prop_type == PropType.BOTH:
+        if attr_type == PropType.SERVER or attr_type == PropType.BOTH:
+            list_svr.append(line)
+        if attr_type == PropType.ECL or attr_type == PropType.BOTH:
+            list_ecl.append(line)
     return None
 
 
-def getText(efl, sfl):
+def getText(s_file, e_file, d_file):
```
Comment (prakashcv13) [inline buildutils/attr_parser.py:118]: how about server_file, ecl_file, def_file?

File: buildutils/attr_parser.py [hunk 11]
```diff
@@ -913,25 +340,21 @@ def main(argv):
         print('Cannot Open Server File!')
         sys.exit(1)
 
+    try:
+        d_file = open(DEF_FILE, 'w', encoding='utf-8')
+    except IOError as err:
+        print(str(err))
+        print('Cannot Open Defines File!')
```
Comment (prakashcv13) [inline buildutils/attr_parser.py:347]: would displaying the file name be helpful?

File: src/server/svr_func.c (hunk 4 above)
Comment (subhasisb) [reply src/server/svr_func.c:1656]: actually, i think u are right? SVR Is more prominent and not SRV, i will switch to SVR for better consistency with rest of the code

File: src/resmom/mock_run.c [hunk 12]
```diff
@@ -181,17 +181,15 @@ mock_run_mom_set_use(job *pjob)
...
+		rd[2] = &svr_resc_def[RESC_CPUT];
+		rd[3] = &svr_resc_def[RESC_CPUPERCENT];
 		rd[4] = NULL;
 	}
-	if (vmemd == NULL) {
-		vmemd = find_resc_def(svr_resc_def, "vmem", svr_resc_size);
-		assert(vmemd != NULL);
-	}
 
+	vmemd = &svr_resc_def[RESC_VMEM];
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/resmom/mock_run.c:191]: actually not required since we are indexing into a pre-known index within a known array, so this is hardcoded.

File: src/server/hook_func.c [hunk 13]
```diff
@@ -3137,16 +3137,14 @@ set_job_reslist(job *pjob, char *hook_name, char *msg, int msg_len,
...
-		if (presc && (presc->rs_value.at_flags & ATR_VFLAG_DEFLT)) {
-			/* changing Resource_List and select is a default */
-			/* clear "select" so it is rebuilt in set_resc_deflt */
-			pseldef->rs_free(&presc->rs_value);
-		}
-		(void)set_resc_deflt((void *)pjob, JOB_OBJECT, NULL);
-	}
+	pseldef = &svr_resc_def[RESC_SELECT];
+	presc = find_resc_entry(jb, pseldef);
+	if (presc && (presc->rs_value.at_flags & ATR_VFLAG_DEFLT)) {
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/server/hook_func.c:3142]: I would like to leave the braces here. Even though there is only one line, there are couple of comments and it can be visually confusing if not within braces

File: src/lib/Libattr/master_job_attr_def.xml [hunk 14]
```diff
@@ -105,1870 +103,1454 @@
...
-      </member_at_flags>
-      <member_at_type>
-         <both>ATR_TYPE_STR</both>
-      </member_at_type>
+      <member_at_action>
+
+#ifdef PBS_MOM
+      NULL_FUNC
+#else
+		ck_chkpnt
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libattr/master_job_attr_def.xml:270]: The NULL_FUNC had spaces, converted it to tabs

File: src/lib/Libattr/master_svr_attr_def.xml (hunk 8 above)
Comment (subhasisb) [reply src/lib/Libattr/master_svr_attr_def.xml:282]: aligned all the xmls using a xml formatter online

File: buildutils/attr_parser.py (hunk 9 above)
Comment (subhasisb) [reply buildutils/attr_parser.py:98]: The parser is hardly worth spending any time documenting. The code is small, self explanatory, and is seldom updated

File: buildutils/attr_parser.py (hunk 10 above)
Comment (subhasisb) [reply buildutils/attr_parser.py:118]: just variables, see the caller - should explain everything

File: buildutils/attr_parser.py (hunk 11 above)
Comment (subhasisb) [reply buildutils/attr_parser.py:347]: yeah that can be changed

**Comment by subhasisb** [comment]
@prakashcv13 have pushed changes, please check further

File: src/lib/Libpython/pbs_python_svr_internal.c [hunk 15]
```diff
@@ -5230,8 +5230,8 @@ _pbs_python_event_set(unsigned int hook_event, char *req_user, char *req_host,
 	}
 
 	lval = min_restart_interval;
-	if (server.sv_attr[(int)SRV_ATR_PythonRestartMinInterval].at_flags & ATR_VFLAG_SET)
-		min_restart_interval = server.sv_attr[(int)SRV_ATR_PythonRestartMinInterval].at_val.at_long;
+	if (server.sv_attr[(int)SVR_ATR_PythonRestartMinInterval].at_flags & ATR_VFLAG_SET)
```
Comment (hirenvadalia) [inline src/lib/Libpython/pbs_python_svr_internal.c:5233]: Optional: do we really need `(int)` type cast? someplace we use, someplace we don't

Comment (subhasisb) [reply src/lib/Libpython/pbs_python_svr_internal.c:5233]: That would be for another PR

**Comment by subhasisb** [comment]
Description: Tests from given sources on platforms CENTOS7, UBUNTU1804, SUSE15, UBUNTU1604
Platforms: CENTOS7,UBUNTU1804,SUSE15,UBUNTU1604
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|2461|3510|0|0|0|0|3510|

**Comment by bhroam** [comment]
@subhasisb I know this has already merged, but can you update the .gitignore file with your newly generated header files?

**Comment by subhasisb** [comment]
@bhroam thanks for pointing out, i will update the .gitignore
